### ISSUE TITLE
NPM packages for vscode-jsonrpc and vscode-languageserver do not include Thenable interface

### ISSUE THREAD

**Issue by felixfbecker** [Microsoft/vscode-languageserver-node#156]
The `Thenable` interface is just placed in a file and not exported. It is used inside files as a global interface, but the file is not imported anywhere, so the distributed package does not include it. It only works in VS Code because VS Code defines the interface itself.

Since the interface is exactly the same as `PromiseLike`, I suggest to simply use `PromiseLike` everywhere.

```
node_modules/vscode-jsonrpc/lib/main.d.ts(8,67): error TS2304: Cannot find name 'Thenable'.
node_modules/vscode-jsonrpc/lib/main.d.ts(46,93): error TS2304: Cannot find name 'Thenable'.
node_modules/vscode-languageserver/lib/files.d.ts(2,83): error TS2304: Cannot find name 'Thenable'.
node_modules/vscode-languageserver/lib/files.d.ts(3,138): error TS2304: Cannot find name 'Thenable'.
node_modules/vscode-languageserver/lib/files.d.ts(5,158): error TS2304: Cannot find name 'Thenable'.
node_modules/vscode-languageserver/lib/files.d.ts(10,155): error TS2304: Cannot find name 'Thenable'.
node_modules/vscode-languageserver/lib/main.d.ts(146,86): error TS2304: Cannot find name 'Thenable'.
node_modules/vscode-languageserver/lib/main.d.ts(153,88): error TS2304: Cannot find name 'Thenable'.
node_modules/vscode-languageserver/lib/main.d.ts(160,92): error TS2304: Cannot find name 'Thenable'.
node_modules/vscode-languageserver/lib/main.d.ts(219,67): error TS2304: Cannot find name 'Thenable'.
```

**Comment by dbaeumer** [comment]
At least the client library is meant to work in VS Code only (so far). Since VS Code uses Thenable we decided for the libraries that are primarily designed to be used with VS Code to use Thanable as well. Otherwise things get too confusing for most of the people. I would rather opt to make the libaray work with Thenable in a non VS Code setup.

**Comment by dbaeumer** [comment]
I tested this with the latest version of the library and they compile fine for me. Setup:

- install vscode-languageserver
- install `@types/node`
- tsconfig.json
```
{
	"compilerOptions": {
		"module": "commonjs",
		"moduleResolution": "node",
		"target": "es6"
	}
}
```
- tsc version 2.1.5
- test.ts
```ts
import * as server from 'vscode-languageserver';

let connection = server.createConnection();
```
- Compile using tsc -p .

Produces no errors for me. @felixfbecker can you please provide a setup in which this result in compile errors for you.

**Comment by felixfbecker** [comment]
I think it is actually more confusing if the interface uses a different name than the standard TS library, and VS Code would still accept the interface since they are equivalent.
But I don't care much, if `Thenable` is included that would work too.

We use the two libraries in the TS/JS language server and previously added a dependency to VS Code which we want to remove.
If you clone Alex' fork https://github.com/alexsaveliev/javascript-typescript-langserver and checkout the branch `no-vscode`, then delete `src/thenable.ts`, and try to compile, you get the above error. This is using version ^2 of the library, I will try latest now.

**Comment by felixfbecker** [comment]
Okay, seems like in v3 vscode-languageserver has `Thenable` included, json-rpc however does not. Unfortunately upgrading to v3 resulted in a lot of other errors since there were breaking changes. v2 does not have `Thenable` included.
I would be fine with working around this by including `Thenable` on our side until we find the time to upgrade to v3. I think json-rpc still needs it though.

**Comment by dbaeumer** [comment]
I am aware of the problem with the 2.x version of the libraries especially when using TS 2.x version. I needed to tweak some things to get this to work nicely which are only available in 3.0.x.

Adopting to the breaking changes is straight forwards. I ask some other users of the library to do so and they told me it was less than 30 minutes. If you have questions about this feel free to ping. I also listed them here: https://github.com/Microsoft/vscode-languageserver-node#breaking-changes

Any objection to close the issue?
